typing/typedecl.ml

```diff
@@ -1103,32 +1103,35 @@ let is_hash id =
   String.length s > 0 && s.[0] = '#'
 
 let marked_as_immediate decl =
-  Builtin_attributes.immediate decl.type_attributes
+  Builtin_attributes.type_repr decl.type_attributes
 
 let compute_immediacy env tdecl =
+  let from_pointer b = if b then Repr_any else Repr_immediate in
   match (tdecl.type_kind, tdecl.type_manifest) with
   | (Type_variant [{cd_args = Cstr_tuple [arg]; _}], _)
     | (Type_variant [{cd_args = Cstr_record [{ld_type = arg; _}]; _}], _)
     | (Type_record ([{ld_type = arg; _}], _), _)
   when tdecl.type_unboxed.unboxed ->
     begin match get_unboxed_type_representation env arg with
-    | Some argrepr -> not (Ctype.maybe_pointer_type env argrepr)
-    | None -> false
+    | Some argrepr -> from_pointer (Ctype.maybe_pointer_type env argrepr)
+    | None -> Repr_any
     end
   | (Type_variant (_ :: _ as cstrs), _) ->
-    not (List.exists (fun c -> c.Types.cd_args <> Types.Cstr_tuple []) cstrs)
-  | (Type_abstract, Some(typ)) ->
-    not (Ctype.maybe_pointer_type env typ)
+    if List.exists (fun c -> c.Types.cd_args <> Types.Cstr_tuple []) cstrs then
+      Repr_address
+    else
+      Repr_immediate
+  | (Type_abstract, Some(typ)) -> Ctype.get_type_repr env typ
   | (Type_abstract, None) -> marked_as_immediate tdecl
-  | _ -> false
+  | _ -> Repr_any
```

Comment:
Couldn't this be `Repr_address` in some cases?
What about records whose fields are not floats?

Perhaps it just doesn't matter, but it's not obvious at quick glance that it doesn't.